ibis/common/deferred.py

```diff
@@ -0,0 +1,608 @@
+
+    def __ne__(self, other: Any) -> DeferredProxy:  # type: ignore
+        return DeferredProxy(BinaryOperator(operator.ne, self, other))
+
+    def __lt__(self, other: Any) -> DeferredProxy:
+        return DeferredProxy(BinaryOperator(operator.lt, self, other))
+
+    def __le__(self, other: Any) -> DeferredProxy:
+        return DeferredProxy(BinaryOperator(operator.le, self, other))
+
+    def __gt__(self, other: Any) -> DeferredProxy:
+        return DeferredProxy(BinaryOperator(operator.gt, self, other))
+
+    def __ge__(self, other: Any) -> DeferredProxy:
+        return DeferredProxy(BinaryOperator(operator.ge, self, other))
+
+    def __and__(self, other: Any) -> DeferredProxy:
+        return DeferredProxy(BinaryOperator(operator.and_, self, other))
+
+    def __rand__(self, other: Any) -> DeferredProxy:
+        return DeferredProxy(BinaryOperator(operator.and_, other, self))
+
+    def __or__(self, other: Any) -> DeferredProxy:
+        return DeferredProxy(BinaryOperator(operator.or_, self, other))
+
+    def __ror__(self, other: Any) -> DeferredProxy:
+        return DeferredProxy(BinaryOperator(operator.or_, other, self))
+
+    def __xor__(self, other: Any) -> DeferredProxy:
+        return DeferredProxy(BinaryOperator(operator.xor, self, other))
+
+    def __rxor__(self, other: Any) -> DeferredProxy:
+        return DeferredProxy(BinaryOperator(operator.xor, other, self))
+
+
+class Variable(FrozenSlotted, Deferred):
+    """Retrieve a value from the context.
+
+    Parameters
+    ----------
+    name
+        The key to retrieve from the state.
+    """
+
+    __slots__ = ("name",)
+    name: Any
+
+    def __init__(self, name):
+        super().__init__(name=name)
+
+    def __repr__(self):
+        return str(self.name)
+
+    def resolve(self, context):
+        return context[self.name]
+
+
+class Just(FrozenSlotted, Deferred):
+    """Construct exactly the given value.
+
+    Parameters
+    ----------
+    value
+        The value to return when the deferred is called.
+    """
+
+    __slots__ = ("value",)
+    value: Any
+
+    @classmethod
+    def __create__(cls, value):
+        if isinstance(value, cls):
+            return value
+        elif isinstance(value, (DeferredProxy, Deferred)):
+            raise TypeError(f"{value} cannot be used as a Just value")
+        elif isinstance(value, collections.abc.Hashable):
+            return super().__create__(value)
+        else:
+            return JustUnhashable(value)
+
+    def __init__(self, value):
+        super().__init__(value=value)
+
+    def __repr__(self):
+        obj = self.value
+        if hasattr(obj, "__deferred_repr__"):
+            return obj.__deferred_repr__()
+        elif callable(obj):
+            return getattr(obj, "__name__", repr(obj))
+        else:
+            return repr(obj)
+
+    def resolve(self, context):
+        return self.value
+
+
+class JustUnhashable(FrozenSlotted, Deferred):
```

Comment:
Possible code simplification:

Could we unify this implementation with the above to only have a single `Just` class that may optionally wrap `value` in `PseudoHashable` if needed? I think the only difference would be the implementation of `.resolve` as:

```python
def resolve(self, context):
    return self.value.obj if isinstance(self.value, PseudoHashable) else self.value
```

Having one fewer wrapped class to worry about might be nice, up to you.

Reply:
They are indeed almost identical, so your request totally makes sense. On the other hand I'd rather keep them separate for performance reasons. The `Just` deferreds are basically the "leaf" objects responsible to produce the requested values and they are frequently used. Perhaps it doesn't make a noticeable performance difference, but mutable values should be rarely used.